compiler/noirc_evaluator/src/acir/types.rs

```diff
@@ -156,11 +157,11 @@ impl AcirValue {
 
     pub(super) fn flat_numeric_types(self) -> Vec<NumericType> {
         match self {
-            AcirValue::Array(_) => {
-                self.flatten().into_iter().map(|(_, typ)| typ.to_numeric_type()).collect()
+            AcirValue::Array(array) => {
+                array.into_iter().flat_map(|elem| elem.flat_numeric_types()).collect()
             }
             AcirValue::DynamicArray(AcirDynamicArray { value_types, .. }) => value_types,
-            _ => unreachable!("An AcirValue::Var cannot be used as an array value"),
```

Comment:
@TomAFrench comments such as these is why I think it would be great to get a walkthrough of how ACIR arrays are supposed to work in general. cc @asterite

Reply:
Yeah they could use a revamp and improved documentation in general. I'll be going over the arrays soon for audit prep work.